bdns/servers.go

```diff
@@ -113,23 +119,68 @@ type dynamicProvider struct {
 	updateCounter *prometheus.CounterVec
 }
 
+// resolveDNSAuthority resolves the DNS authority to use for resolution of other
+// DNS backends. The DNS authority can be specified as a hostname or IP address,
+// with or without a port. If the authority is specified as a hostname it will
+// be resolved via the system DNS. If the authority is specified as an IP
+// address it will be used directly, defaulting to port 53 if no port is
+// specified.
+func resolveDNSAuthority(d string) (string, error) {
+	host, port, err := net.SplitHostPort(d)
+	if err != nil {
+		// Assume host with no port specified, default port to 53.
+		host = d
+		port = "53"
+
+	}
+	if net.ParseIP(host) != nil {
+		return net.JoinHostPort(host, port), nil
+	}
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return "", fmt.Errorf("during A/AAAA lookup of %q: %s", d, err)
+	}
+	if len(ips) <= 0 {
```

Comment:
This actually crossed my mind. It's typically assumed that the caller will take the first result because the results of an A/ AAAA lookup have no guaranteed order, this is sometimes used a kind of cheap round-robin load balancing.